src/btree/bt_cursor.c

```diff
@@ -674,23 +720,83 @@ int
 __wt_btcur_remove(WT_CURSOR_BTREE *cbt)
 {
 	WT_BTREE *btree;
+	WT_CURFILE_OP_DECL;
 	WT_CURSOR *cursor;
 	WT_DECL_RET;
 	WT_SESSION_IMPL *session;
+	bool positioned;
 
 	btree = cbt->btree;
 	cursor = &cbt->iface;
 	session = (WT_SESSION_IMPL *)cursor->session;
 
+	WT_CURFILE_OP_PUSH;
+
 	WT_STAT_CONN_INCR(session, cursor_remove);
 	WT_STAT_DATA_INCR(session, cursor_remove);
 	WT_STAT_DATA_INCRV(session, cursor_remove_bytes, cursor->key.size);
 
-retry:	WT_RET(__cursor_func_init(cbt, true));
+	/*
+	 * WT_CURSOR.remove has a unique semantic, the cursor stays positioned
+	 * if it starts positioned, otherwise clear the cursor on completion.
+	 */
+	positioned = F_ISSET(cursor, WT_CURSTD_KEY_INT);
 
-	switch (btree->type) {
-	case BTREE_COL_FIX:
-	case BTREE_COL_VAR:
+retry:
+	/*
+	 * If removing with overwrite configured, and positioned to an on-page
+	 * key, the update doesn't require another search. The cursor won't be
+	 * positioned on a page with an external key set, but be sure.
+	 */
+	if (WT_PAGE_PINNED(cbt) &&
+	    F_ISSET_ALL(cursor, WT_CURSTD_KEY_INT | WT_CURSTD_OVERWRITE)) {
+		WT_ERR(__wt_txn_autocommit_check(session));
+
+		/*
+		 * The cursor position may not be exact (the cursor's comparison
+		 * value not equal to zero). Correct to an exact match so we can
+		 * remove whatever we're pointing at.
+		 */
+		cbt->compare = 0;
```

Comment:
Surprised me, too. :)

Yes, I believe this change is necessary in order to do a `search_near` followed by a remove, although that's not where I saw the failure.

We generally don't reset `cbt->compare`, it can be left to non-zero after any search (for example, a search that failed to find a searched-for key, or an insert of a new key into the tree).

So, for example, if you insert a new key, then re-use the cursor for iteration, and then call remove, I think it will hit this problem.

I thought about changing things so every cursor operation leaves `cbt->compare` in a known state, but that will be messy and more fragile than this change.

Another way to think about it: Remove can skip the search if `WT_CURSTD_KEY_INT` is set, but we still have to set `cbt->compare` for the modify functions as the search function would have done, had we called it.